WalletWasabi/Blockchain/Analysis/FeesEstimation/FeeProviders.cs

```diff
@@ -49,20 +36,23 @@ private void SetAllFeeEstimate()
 		{
 			lock (Lock)
 			{
+				AllFeeEstimate? feeEstimateToSet = null;
 				IFeeProvider[] providerArray = Providers.ToArray();
```

Comment:
Why is this being done? Why not simple foreach?

Reply:
The last provider is always the backend (providerArray.Length - 1)

Reply:
Hm. That's hardly obvious. The code would be simpler if:

```csharp
public FeeProviders(IEnumerable<IFeeProvider> feeProviders)
```
->
```csharp
public FeeProviders(List<IFeeProvider> feeProviders)
```

It would be probably logically clearer to have:

```csharp
public FeeProviders(List<IFeeProvider> feeProviders, IFeeProvider backup)
```

and the code would be much shorter and clearer.

Reply:
Then it would be:

```csharp
using System;
using System.Collections.Generic;

namespace WalletWasabi.Blockchain.Analysis.FeesEstimation
{
	public class FeeProviders : IFeeProvider, IDisposable
	{
		public FeeProviders(List<IFeeProvider> feeProviders, IFeeProvider baseProvider)
		{
			Providers = feeProviders;
			BaseProvider = baseProvider;
			Lock = new object();

			SetAllFeeEstimate();

			foreach (var provider in Providers)
			{
				provider.AllFeeEstimateChanged += Provider_AllFeeEstimateChanged;
			}
		}

		public event EventHandler<AllFeeEstimate>? AllFeeEstimateChanged;

		public AllFeeEstimate AllFeeEstimate { get; private set; }

		private List<IFeeProvider> Providers { get; }
		private IFeeProvider BaseProvider { get; }
		private object Lock { get; }

		private void Provider_AllFeeEstimateChanged(object? sender, AllFeeEstimate e)
		{
			SetAllFeeEstimate();
		}

		private void SetAllFeeEstimate()
		{
			lock (Lock)
			{
				AllFeeEstimate? feeEstimateToSet = null;

				foreach (IFeeProvider provider in Providers)
				{
					if (provider.AllFeeEstimate is { IsAccurate: bool isAccurate } af && isAccurate)
					{
						feeEstimateToSet = af;
						break;
					}
				}

				AllFeeEstimate = feeEstimateToSet ?? BaseProvider.AllFeeEstimate;
			}

			AllFeeEstimateChanged?.Invoke(this, AllFeeEstimate);
		}

		#region IDisposable Support

		private volatile bool _disposedValue = false; // To detect redundant calls

		protected virtual void Dispose(bool disposing)
		{
			if (!_disposedValue)
			{
				if (disposing)
				{
					foreach (var provider in Providers)
					{
						provider.AllFeeEstimateChanged -= Provider_AllFeeEstimateChanged;
					}
				}

				_disposedValue = true;
			}
		}

		// This code added to correctly implement the disposable pattern.
		public void Dispose()
		{
			// Do not change this code. Put cleanup code in Dispose(bool disposing) above.
			Dispose(true);
		}

		#endregion IDisposable Support
	}
}
```

Reply:
The code still has one nullable warning - `AllFeeEstimate`. I'm not sure whether it can be null. But it seems that providers can also return null. But I'm not sure.

Reply:
I don't want to take care of Nullable things here. That would even change the interface (IFeeProvider) at the end.

Reply:
Hm, but then nobody will do it. :-|